### PR TITLE
Support go-to-definition for imports of arbitrary files

### LINKED ISSUE #41861: Go-to-definition doesn't work when a file exists but is not a module
```ts
// @filename: scriptThing.ts
console.log("woooo side effects")

// @filename: stylez.css
div {
  color: magenta;
}

// @filename: moduleThing.ts

// not a module, but we should let you jump to it.
import /*1*/"./scriptThing";

// not JS/TS, but if we can, you should be able to jump to it.
import /*2*/"./stylez.css";
```

Go to definition on `/*1*/` and `/*2*/` should work!

### PR DESCRIPTION
**Pull request by andrewbranch**
Fixes #41861

The first commit adds support for jumping to arbitrary files from import/require module specifiers, whether those are scripts or non-source files like CSS (thanks @DanielRosenwasser for the idea that this should work).

The second commit goes further by allowing us to return a successful response for files that _don’t exist_. I was curious what would happen, so I removed the `host.fileExists` check and made sure we don’t try to read a file’s text to convert `{ position: 0 }` to `{ line: 0, column: 0 }`. The result in VS Code was this:

File does not exist, do you want to create it?

Having the quick option to create the file seemed like a win to me; better than doing nothing. But I’m open to debate on whether we want that.

Also, if anyone knows how this will work in Visual Studio or has a quick way to test it, that would be appreciated!

### CODE REVIEW

File: src/services/goToDefinition.ts [hunk 1]
```diff
@@ -108,24 +108,41 @@ namespace ts.GoToDefinition {
...
         if (libReferenceDirective) {
             const file = program.getLibFileFromReference(libReferenceDirective);
-            return file && { reference: libReferenceDirective, file };
+            return file && { reference: libReferenceDirective, fileName: file.fileName };
+        }
+
+        if (sourceFile.resolvedModules?.size) {
+            const node = getTokenAtPosition(sourceFile, position);
+            if (isModuleSpecifierLike(node) && !pathIsBareSpecifier(node.text) && sourceFile.resolvedModules.has(node.text)) {
+                const fileName = sourceFile.resolvedModules.get(node.text)?.resolvedFileName
```
Comment (sheetalkamat) [inline src/services/goToDefinition.ts:134]: i dont think we should do that.. what with option of using `rootDirs` to determine where this is relative to etc ?

Comment (andrewbranch) [reply src/services/goToDefinition.ts:134]: I assume you’re not talking about this line, but the next?

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this

**Comment by DanielRosenwasser** [comment]
Heads up @mjbvz and @minestarks.

**Comment by DanielRosenwasser** [comment]
How does this work with partial semantic mode (including in the browser)?

What will happen if we try go-to-definition on `thing` in this example?

```ts
import { thing } from "./vandelay.js";

thing/**/();
```

**Comment by andrewbranch** [comment]
> What will happen if we try go-to-definition on `thing`

Whatever the current behavior for go-to-definition on the identifier reference, that will be unchanged.

> or the module specifier

Don’t we already resolve one level of direct imports from your current source file? That should be unchanged as well. If we don’t, we would return a response saying the definition is at position 0 of `./thing.js`, resolved relative to whatever we said the path of the current source file is. If we were to re-add the `host.fileExists` check and we haven’t already successfully resolved and loaded `./thing.js` for normal module resolution reasons, we would not offer a definition there.

File: src/services/goToDefinition.ts (hunk 1 above)
Comment (andrewbranch) [reply src/services/goToDefinition.ts:134]: The relative path resolution will only kick in if _real_ module resolution failed. This applies only to file types we don’t understand, like CSS, and to files that don’t exist, so you have a quick shortcut to create it. So yes, it’s a best guess, but I think it’s a pretty good guess (I don’t think very many people use `rootDirs`), and it replaces a total refusal to try.

**Comment by DanielRosenwasser** [comment]
I mostly ask because #39037 is a related issue I've been interested in unlocking as well, but I get that they work differently.

**Comment by sandersn** [comment]
Answering a question that *nobody* has asked, I tested this with emacs. This gives a picture of how a simply written language plugin will behave and is usually similar to how VS works too. 

Everything works fine. For the non-existent file, the editor raises an error inside `tide-get-file-buffer` and since the error's not handled, it's displayed to the user. That's a fine user experience, and I would be surprised if it's worse in other editors.

File: src/services/goToDefinition.ts (hunk 1 above)
Comment (sheetalkamat) [reply src/services/goToDefinition.ts:134]: Yeah i meant next line... 
 wonder if these should be flagged as best guess or some such flag instead..
as @DanielRosenwasser pointed, we dont resolve imports currently in partial semantic mode, so adding a test that it works for that would be good too..

File: src/services/goToDefinition.ts [hunk 2]
```diff
@@ -108,24 +108,41 @@ namespace ts.GoToDefinition {
...
         const libReferenceDirective = findReferenceInPosition(sourceFile.libReferenceDirectives, position);
         if (libReferenceDirective) {
             const file = program.getLibFileFromReference(libReferenceDirective);
-            return file && { reference: libReferenceDirective, file };
+            return file && { reference: libReferenceDirective, fileName: file.fileName, file };
+        }
+
+        if (sourceFile.resolvedModules?.size) {
+            const node = getTokenAtPosition(sourceFile, position);
+            if (isModuleSpecifierLike(node) && !pathIsBareSpecifier(node.text) && sourceFile.resolvedModules.has(node.text)) {
```
Comment (sheetalkamat) [inline src/services/goToDefinition.ts:133]: isExternalModuleNameRelative is better readable here ?

Comment (andrewbranch) [reply src/services/goToDefinition.ts:133]: TIL those functions are in fact inverses 👍

Comment (andrewbranch) [reply src/services/goToDefinition.ts:133]:
```suggestion
            if (isModuleSpecifierLike(node) && isExternalModuleNameRelative(node.text) && sourceFile.resolvedModules.has(node.text)) {
```

**Comment by andrewbranch** [comment]
> we dont resolve imports currently in partial semantic mode, so adding a test that it works for that would be good too

@sheetalkamat I’m looking at adding a case in partialSemanticServer.ts now, and it seems imports are resolved... the source file’s `resolvedModules` contains correct resolutions.

**Comment by sheetalkamat** [comment]
@andrewbranch interesting.. till you pointed out,  I didn't realize we were doing module resolution as part of `noResolve`. PR #4368 added this. Not sure if this additional cost is really what we want with partial semantic server or not but that can be handled separately.

**Comment by andrewbranch** [comment]
In that case, this feature currently works exactly the same for partial semantic mode as full semantic mode, and the new test will ensure that we don’t break it if we change partial semantic mode later.

File: src/services/goToDefinition.ts (hunk 1 above)
Comment (sheetalkamat) [reply src/services/goToDefinition.ts:134]: Still believe we probably would want to flag guesses in definition so vscode can potentially show some kind of message or something like it does for completions ?

File: src/services/goToDefinition.ts [hunk 3]
```diff
@@ -108,24 +108,41 @@ namespace ts.GoToDefinition {
...
             const file = program.getLibFileFromReference(libReferenceDirective);
-            return file && { reference: libReferenceDirective, file };
+            return file && { reference: libReferenceDirective, fileName: file.fileName, file };
+        }
+
+        if (sourceFile.resolvedModules?.size) {
+            const node = getTokenAtPosition(sourceFile, position);
+            if (isModuleSpecifierLike(node) && isExternalModuleNameRelative(node.text) && sourceFile.resolvedModules.has(node.text)) {
+                const fileName = sourceFile.resolvedModules.get(node.text)?.resolvedFileName
+                    || resolvePath(getDirectoryPath(sourceFile.fileName), node.text);
```
Comment (sheetalkamat) [inline src/services/goToDefinition.ts:135]: What about this marked as a guess? Do we need that?

File: src/services/goToDefinition.ts (hunk 1 above)
Comment (andrewbranch) [reply src/services/goToDefinition.ts:134]: @DanielRosenwasser @mjbvz thoughts? We can discuss at the editor meeting tomorrow if needed.

**Comment by andrewbranch** [comment]
New changes:

1. Whenever the file reference lookup does not yield an actual source file (this could be because the file doesn’t exist, _or_ because it’s a file type like CSS that the compiler doesn’t care about), a new `unverified: true` property is added to the definition in the response. VS Code could use this to show a less-error-y prompt to create a new file if it finds that the file doesn’t exist.
2. Whenever (1) happens, instead of immediately returning that speculative file result, we continue other methods of definition lookup and combine the results. This specifically allows us to return two definitions for pattern ambient module matches: one for the actual file referenced, and one for the ambient module declaration. Screen cap:

   Go to definition on index.css returns a result for the CSS file and for a "*.css" pattern ambient module declaration

   Notes: I’m not sure why the second definition is expanded by default instead of the first. Possibly because the first has a zero-length definition span? Also, the underlining on only part of the path seems to be unrelated to TypeScript’s response (the span we send there is the whole module specifier, quotes included).

**Review by RyanCavanaugh — APPROVED** [review]

File: src/services/goToDefinition.ts [hunk 4]
```diff
@@ -2,8 +2,10 @@
 namespace ts.GoToDefinition {
     export function getDefinitionAtPosition(program: Program, sourceFile: SourceFile, position: number): readonly DefinitionInfo[] | undefined {
         const resolvedRef = getReferenceAtPosition(sourceFile, position, program);
-        if (resolvedRef) {
-            return [getDefinitionInfoForFileReference(resolvedRef.reference.fileName, resolvedRef.file.fileName)];
+        const fileReferenceDefinition = resolvedRef && [getDefinitionInfoForFileReference(resolvedRef.reference.fileName, resolvedRef.fileName, !!resolvedRef.file)] || emptyArray;
```
Comment (sheetalkamat) [inline src/services/goToDefinition.ts:5]: Having source file present is not good enough indicator of "guess" as `resolvePath(getDirectoryPath(sourceFile.fileName), node.text);` this is just a guess. so wether file exists or not it should be marked as unverified?

**Review by sheetalkamat — CHANGES_REQUESTED** [review]

**Review by sheetalkamat — APPROVED** [review]

File: src/services/goToDefinition.ts [hunk 5]
```diff
@@ -111,24 +113,42 @@ namespace ts.GoToDefinition {
...
+                const fileName = verifiedFileName || resolvePath(getDirectoryPath(sourceFile.fileName), node.text);
+                return {
+                    file: program.getSourceFile(fileName),
+                    fileName,
+                    reference: {
+                        pos: node.getStart(),
+                        end: node.getEnd(),
+                        fileName: node.text
+                    },
+                    unverified: !!verifiedFileName,
```
Comment (sheetalkamat) [inline src/services/goToDefinition.ts:149]: @andrewbranch just realised that we are not passing this on in the through server even though we added this property to protocol.
[tsserver37220.log](https://github.com/microsoft/TypeScript/files/6240528/tsserver37220.log)

**Pull request merged**